### ISSUE TITLE
Cannot use ftp protocol in install.sh for downloading non-existent packages in secure environments

### ISSUE THREAD

**Issue by singleterry** [sourceryinstitute/OpenCoarrays#585]
- OpenCoarrays Version: <2.2.0>
- Fortran Compiler: <gcc 7.3.0>
- C compiler used for building lib: <ggc 7.3.0>
- Installation method: <install.sh>
- Output of `uname -a`: Linux hze.larc.nasa.gov 2.6.32-431.11.2.el6.x86_64 #1 SMP Tue Mar 25 19:59:55 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux
- MPI library being used: <3.2>
- Machine architecture and number of physical cores: not relevant
- Version of CMake: <3.4.0>

When using the install.sh script, the prerequisites downloaded with the ftp protocol will fail in secure environments.  The protocol used to download those files should be changed.

**Comment by singleterry** [comment]
Assume I can just reply back!?!

The ftp protocol is forbidden in secure environments.  Therefore when curl (and I assume wget) tried to access ftp://url, it cannot find the site because the ftp protocol is blocked.  Of course, when I http://url, it finds it fine, I download what it was trying to download and rerun the install script with success.  Maybe change to https:// or sftp://

This did not happen with all of the curl/wget calls.  Some were to https:// and they worked fine!

I hope that helps!?!

Robert

*******************************************************************************
Robert C. Singleterry Jr., Ph.D.
Research Engineer and NASA Administrator’s Fellow (Cohort 6: 2002-2004)
Durability, Damage Tolerance, & Reliability Branch
Langley Research Center, National Aeronautics and Space Administration
Building 1205, Room 285, Mail Stop 188E
2 West Reid Street
Hampton, VA 23681-2199
Work Voice: (757) 864-1437  FAX: (757) 864-4385   Personal Cell: (757) 371-4848
*******************************************************************************

**Comment by zbeekman** [comment]
Yes, email replies get forwarded. 🙂

Thanks for clarification. We certainly should review all of our download URLs and favor encrypted variants wherever possible via `https` and, if/where needed, `sftp`. Sometimes, however, people are on systems lacking CA certs, and curl or wget fail with SSL errors, but I’d rather someone not be able to install OC, than be the vector used to compromise a machine through a MITM software supply chain attack. 

I’ll assign this issue to myself to ensure I review the urls and update to use encryption wherever possible.

**Comment by singleterry** [comment]
Hello,

Maybe make that a question to ask the user??

Robert

*******************************************************************************
Robert C. Singleterry Jr., Ph.D.
Research Engineer and NASA Administrator’s Fellow (Cohort 6: 2002-2004)
Durability, Damage Tolerance, & Reliability Branch
Langley Research Center, National Aeronautics and Space Administration
Building 1205, Room 285, Mail Stop 188E
2 West Reid Street
Hampton, VA 23681-2199
Work Voice: (757) 864-1437  FAX: (757) 864-4385   Personal Cell: (757) 371-4848
*******************************************************************************

**Comment by rouson** [comment]
@singleterry  Thanks for submitting this. 

@zbeekman The URL for each package the installer knows how to install is listed [here].  The only ones that use the `ftp` protocol are the ones on the GNU ftp server `ftp.gnu.org`.  If there's a reliable mirror that uses something other than `ftp`, the only change that is needed is to change each corresponding URL in the aforementioned file. Most likely, everything will work from there.

In case it helps, if you ever want to see what URL the installer will use for a given package, pass the package name with the`-U` argument:
```
$ ./install.sh -U gcc
ftp://ftp.gnu.org:/gnu/gcc/gcc-8.2.0/gcc-8.2.0.tar.gz
```

[here]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/prerequisites/build-functions/set_or_print_url.sh#L29

**Comment by rouson** [comment]
I see there's a GNU mirror list here: https://www.gnu.org/prep/ftp.en.html.

**Comment by zbeekman** [comment]
@rouson I think we want: https://ftpmirror.gnu.org/gcc/gcc-8.2.0/gcc-8.2.0.tar.xz

I *think* this redirects to a suitable mirror that is closest to the user (and has https/ssl encryption) so it should be fast.

**Comment by rouson** [comment]
I'll test the switch and push a fix.

**Comment by rouson** [comment]
This is now fixed in PR #594 and will appear in the next release.
